### ISSUE TITLE
Abstract over JSON library

### ISSUE THREAD

**Issue by LMnet** [lihaoyi/mill#949]
Hi! I want to discuss the possibility to abstract over JSON library in the mill.

At the current moment, mill uses an upickle library for serializing/deserializing tasks metadata. The serialization format for this metadata is JSON. And, as far as I understand, mill doesn't use any specific upickle functionality. It looks like any JSON library could be used for that. 

I have nothing against upickle, but I think this hard binding could be problematic for external libraries integration. For example, I wanted to use the [skuber](https://github.com/doriordan/skuber) library to describe Kubernetes entities in my build files and deploy them from the mill. Something like this (simplified example):

```scala
trait Deploy extends Module {
  def deployment: Target[Deployment]
  def deploy = ??? // deploy logic here
}
```
But, if I want to describe `Deployment` in my build, I have to be able to serialize it, because it is `Target`. This means that I have to create an upickle `ReadWriter[Deployment]`. It's doable, but `Deployment` is a pretty complex type and contains a lot of other skuber entities. So, in practice, I have to describe `ReadWriter` instances for almost all skuber entities. This is tedious and error-prone.

Fortunately, skuber provides play-json instances for all its entities. And, if mill will have the possibility to integrate with any JSON library, I could just provide a bridge from play-json to mill JSON. Moreover, these bridges could be provided as libraries for a few most popular JSON libraries (circe, play-json, spray?). With ready-made JSON integrations, integration with skuber will be very simple.

So, I propose to create an intermediate JSON layer in the mill. It will support upickle by default. So, all current instances will not be touched. But, in the case of necessity, integrations with other JSON libraries could be made.

To test this idea I created a [WIP pull request](https://github.com/lihaoyi/mill/pull/948/files). Implementation is not finished, but it looks like needed changes are not massive.

What do you think about it?

**Comment by lihaoyi** [comment]
Would an easier approach be converting creating `ReadWriter[Deployment]` based on your play-json reader/writer for `Deployment`? If you don't mind an intermediate data structure, it should be easy to define a `ReadWriter[Deployment]` as `Deployment <-play-json-reader-writer-> JSON String <-ujson.reformatTo-> meta.json`. A similar approach would work for any other data type with a JSON reader/writer via any other library as well.

This ends up with one intermediate data structure, but would be significantly less intrusive than trying to refactor Mill itself to be json-library-agnostic.

**Comment by LMnet** [comment]
You mean something like this:
```scala
def playJsonToUpickleRW[A](implicit playJsonFormat: Format[A]): ReadWriter[A] = {
  readwriter[String].bimap[A](
    a => playJsonFormat.writes(a).toString(),
    str => playJsonFormat.reads(Json.parse(str)).get,
  )
}
```

But it will lead to double json serialization/deserialization. Maybe in terms of a build tool it's not a big deal. But considering that making mill json library agnostic is not a very complex task (from first sight) why not trying that?

**Comment by LMnet** [comment]
Ok, that sounds reasonable. For now, I will stick with the conversion on the user site.

I want to give you some more details about my proposal. When I faced an issue with play-json for the first time I tried to derive `ReadWriter [A]` instances from the play-json `Format [A]` instances without intermediate `String` layer. And I failed because `ReadWriter` uses` Visitor` and has a pretty specific API, so I didn't find a way to create a conversion. So, I thought that the intermediate mill json layer might be useful in this case.
